### PR TITLE
BUG: Don't ignore na_rep in DataFrame.to_html

### LINKED ISSUE #9046: DataFrame.to_latex() should honor na_rep after formatter.
In the example below I would expect that 'np.nan' would be replaced by 'aaah'.

```
In [3]: df = pd.DataFrame({'a': np.arange(3), 'b': [0.1, np.nan, 0.2]})

In [4]: df
Out[4]: 
   a    b
0  0  0.1
1  1  NaN
2  2  0.2

In [5]: print(df.to_latex(formatters={'b': lambda x: '{:0.4f}'.format(x)}, na_rep='aaah'))
\begin{tabular}{lrr}
\toprule
{} &  a &      b \\
\midrule
0 &  0 & 0.1000 \\
1 &  1 &    nan \\
2 &  2 & 0.2000 \\
\bottomrule
\end{tabular}
```

### LINKED ISSUE #13828: to_html ignores na_rep when float_format set
#### Code Sample, a copy-pastable example if possible

http://stackoverflow.com/questions/23700835/pandas-to-html-no-value-representation

df = pd.DataFrame([['A',1.2225],['A',]], columns=['Group','Data'])
df.to_html(na_rep="Ted",float_format='{0:.2f}'.format)
#### Expected Output

<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Group</th>
      <th>Data</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td> A</td>
      <td>1.22</td>
    </tr>
    <tr>
      <th>1</th>
      <td> A</td>
      <td> Ted</td>
    </tr>
  </tbody>
</table>

## INSTALLED VERSIONS

commit: None
python: 3.5.1.final.0
python-bits: 64
OS: Darwin
OS-release: 15.6.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8

pandas: 0.18.0
nose: 1.3.7
pip: 8.1.2
setuptools: 20.3
Cython: 0.23.4
numpy: 1.10.4
scipy: 0.17.0
statsmodels: 0.6.1
xarray: None
IPython: 4.1.2
sphinx: 1.3.5
patsy: 0.4.0
dateutil: 2.5.1
pytz: 2016.2
blosc: None
bottleneck: 1.0.0
tables: 3.2.2
numexpr: 2.5
matplotlib: 1.5.1
openpyxl: 2.3.2
xlrd: 0.9.4
xlwt: 1.0.0
xlsxwriter: 0.8.4
lxml: 3.6.0
bs4: 4.4.1
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: 1.0.12
pymysql: None
psycopg2: None
jinja2: 2.8
boto: 2.39.0

### PR DESCRIPTION
**Pull request by dsaxton**
- [x] closes #13828
- [x] closes #9046
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by simonjayhawkins** [comment]
Off the top of my head I recall a discussion where this may not be a bug. There may be a duplicate issue.

I've not checked the code. but if a user supplies a formatter, they are responsible for the na values too. Is that what is happening here?

**Comment by dsaxton** [comment]
> Off the top of my head I recall a discussion where this may not be a bug. There may be a duplicate issue.
> 
> I've not checked the code. but if a user supplies a formatter, they are responsible for the na values too. Is that what is happening here?

There was an old PR similar to this that was almost merged but ended up being closed as stale, so I assume this is considered a real bug.

I could see how it could be seen as ambiguous since NaN is technically a float, so that the user is asking for two different formats for these values (but in that case maybe the docs should explicitly clarify what happens in this situation).

**Comment by simonjayhawkins** [comment]
does this also close #9046? and presumably to_string does the same?

**Comment by simonjayhawkins** [comment]
I've not yet found the issue but from

> float_format: one-parameter function, optional, default None
Formatter function to apply to columns’ elements if they are floats. The result of this function must be a unicode string.

so the changes here break existing code (this is a long standing behaviour even if considered a bug)

```
def my_formatter(x):
    if np.isnan(x):
        return "ted"
    else:
        return str(x)


df = pd.DataFrame(
    [
        ["A", 1.2225],
        [
            "A",
        ],
    ],
    columns=["Group", "Data"],
)
df.to_html(float_format=my_formatter)
```

1.1.2
image

this PR
image

**Comment by dsaxton** [comment]
> does this also close #9046? and presumably to_string does the same?

Yes, tests added for those cases

**Comment by dsaxton** [comment]
> so the changes here break existing code (this is a long standing behaviour even if considered a bug)

Ah I was forgetting "NaN" is actually the default and not None. Made a change which fixes this but leaves a different "bug" (albiet a more obscure one, it only ignores the specific string "NaN"):

```python
[ins] In [1]: import numpy as np
         ...: import pandas as pd
         ...:
         ...: def my_formatter(x):
         ...:     if np.isnan(x):
         ...:         return "ted"
         ...:     else:
         ...:         return str(x)
         ...:
         ...:
         ...: df = pd.DataFrame(
         ...:     [
         ...:         ["A", 1.2225],
         ...:         [
         ...:             "A",
         ...:         ],
         ...:     ],
         ...:     columns=["Group", "Data"],
         ...: )
         ...: print(df.to_html(na_rep="NaN", float_format=my_formatter))
         ...:
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Group</th>
      <th>Data</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>A</td>
      <td>1.2225</td>
    </tr>
    <tr>
      <th>1</th>
      <td>A</td>
      <td>ted</td>
    </tr>
  </tbody>
</table>
```

File: pandas/io/formats/format.py [hunk 1]
```diff
@@ -1533,7 +1533,14 @@ def format_values_with(float_format):
     def _format_strings(self) -> List[str]:
         # shortcut
         if self.formatter is not None:
```
Comment (jreback) [inline pandas/io/formats/format.py:1535]: actually this entire section you added should be instead on L1440

Comment (jreback) [reply pandas/io/formats/format.py:1535]: and you will want to do something like
```
mask = isna(values)	
values = np.array(values, dtype="object")
values[mask] = na_rep
imask = (~mask).ravel(
values.flat[imask] = np.array(	
                [formatter(val) for val in values.ravel()[imask]]
            )
```

you will want to factor that out into a function and use it above in 2 (or more places)

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
>K Can add a release note once the doc is started for 1.1.4

generally we won't do older bug fixes on a point release; only regressions. the reason is sometimes bug fixes can introduce a further bug :-> want to minimize the risk.

File: pandas/io/formats/format.py (hunk 1 above)
Comment (dsaxton) [reply pandas/io/formats/format.py:1535]: I'm not sure I understand, `get_result_as_array` never actually gets called here. Are you thinking the formatter should already be handling NaN?

File: doc/source/whatsnew/v1.2.0.rst [hunk 2]
```diff
@@ -337,7 +337,7 @@ Conversion
 Strings
 ^^^^^^^
 - Bug in :meth:`Series.to_string`, :meth:`DataFrame.to_string`, and :meth:`DataFrame.to_latex` adding a leading space when ``index=False`` (:issue:`24980`)
--
+- Bug in :meth:`DataFrame.to_html`, :meth:`DataFrame.to_string`, and :meth:`DataFrame.to_latex` ignoring the ``na_rep`` argument when ``float_format`` was also specified (:issue:`9046`, :issue:`13828`)
```
Comment (simonjayhawkins) [inline doc/source/whatsnew/v1.2.0.rst:340]: I think this goes in the I/O section.

File: pandas/tests/io/formats/test_format.py [hunk 3]
```diff
@@ -3432,3 +3432,14 @@ def test_format_remove_leading_space_dataframe(input_array, expected):
     # GH: 24980
     df = pd.DataFrame(input_array).to_string(index=False)
     assert df == expected
+
+
+@pytest.mark.parametrize("na_rep, string", [("NaN", "nan"), ("Ted", "Ted")])
```
Comment (simonjayhawkins) [inline pandas/tests/io/formats/test_format.py:3437]: i'm not sure about using "nan" when `na_rep` is "NaN", although that is the string output of `"{:.2f}".format(np.nan)`.

Maybe could use `lib.no_default` for `na_rep`, raise if `float_format` specified and use "NaN" if `float_format` not specified.

maybe worth considering passing `np.nan` to `func` if passed to `float_format`, and if result is a string assume `func` handles missing values, something like

```
try:
   func_handles_na = isinstance(func(np.nan), str)
except Execption:
   func_handles_na = False
```

and then use `na_rep` if `func_handles_na` is `False`. but I'm not sure if we gain anything.

The main issue with the custom formatters (applies to `formatters` kwarg as well) is that we do not give complete control to the custom formatter. off the top of my head, strings maybe trimmed, spaces added, precision changed, truncation applied.

The other issue is that the EAs use the custom formatters. So this is not necessarily an easy issue to fix in isolation.

File: pandas/tests/io/formats/test_to_html.py [hunk 4]
```diff
@@ -820,3 +820,38 @@ def test_html_repr_min_rows(datapath, max_rows, min_rows, expected):
...
+      <td>1.22</td>
+    </tr>
+    <tr>
+      <th>1</th>
+      <td>A</td>
+      <td>{string}</td>
+    </tr>
+  </tbody>
+</table>"""
+    assert result == expected
```
Comment (simonjayhawkins) [inline pandas/tests/io/formats/test_to_html.py:857]: for html tests, can use expected_html fixture. see test_to_html_justify for usage as template.

**Comment by simonjayhawkins** [comment]
so returning to issue #13828, the fact that "{:.2f}".format(np.nan)" is "nan", I think the output is correct. and this is not a bug https://github.com/pandas-dev/pandas/pull/36690#issuecomment-700915940

but raising if `na_rep` is specified when `float_format` is specified maybe a better solution.

currently, in this PR

image

This is inconsistent.

**Comment by dsaxton** [comment]
> but raising if `na_rep` is specified when `float_format` is specified maybe a better solution.

Raising when both are specified makes sense to me (or rather when float_format is specified and na_rep != "NaN", unless the default were also changed), since there is in fact no way to apply this consistently

**Comment by simonjayhawkins** [comment]
despite my previous comments, I think we could only raise if not a mixed frame. otherwise may cause another regression.

I looked at sorting out some of the formatting in the past, and it's very easy to break things. (needs many more tests with lots of parameterisation to know for sure)

It is maybe that we do need to break things, and maybe the case in https://github.com/pandas-dev/pandas/pull/36690#issuecomment-700957174 is one of those.

File: pandas/io/formats/format.py (hunk 1 above)
Comment (jreback) [reply pandas/io/formats/format.py:1535]: no my point is to share code; you are doing virtually the same thing, just in another way

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/io/formats/format.py (hunk 1 above)
Comment (dsaxton) [reply pandas/io/formats/format.py]: Ok I see what you mean

**Review by simonjayhawkins — APPROVED** [review]
Thanks @dsaxton lgtm.

regarding the corner case where a user custom function handles missing values which is now broken, any idea on how we should communicate that to the users?

**Comment by dsaxton** [comment]
> Thanks @dsaxton lgtm.
> 
> regarding the corner case where a user custom function handles missing values which is now broken, any idea on how we should communicate that to the users?

Maybe warrants a note in user_guide/io.rst?

**Comment by simonjayhawkins** [comment]
> Maybe warrants a note in user_guide/io.rst?

how about just a versionchanged tag for float_format with a one-liner along the lines "If a function is passed, only non-missing values are passed to the function and na_rep is used for missing values"

**Comment by dsaxton** [comment]
> how about just a versionchanged tag for float_format with a one-liner along the lines "If a function is passed, only non-missing values are passed to the function and na_rep is used for missing values"

Added a note and versionchanged to the common docstring

File: pandas/io/formats/format.py [hunk 5]
```diff
@@ -1444,8 +1449,19 @@ def get_result_as_array(self) -> np.ndarray:
         Returns the float values converted into strings using
         the parameters given at initialisation, as a numpy array
         """
+
+        def format_with_na_rep(values, formatter, na_rep):
```
Comment (jreback) [inline pandas/io/formats/format.py:1453]: can you type

File: pandas/io/formats/format.py [hunk 6]
```diff
@@ -1444,8 +1449,19 @@ def get_result_as_array(self) -> np.ndarray:
...
+            mask = isna(values)
+            formatted = np.array(
+                [
```
Comment (jreback) [inline pandas/io/formats/format.py:1456]: can you use this function more generally? (e.g. maybe define it in the module); can be a followup as well.

**Review by jreback — CHANGES_REQUESTED** [review]
minor comment, ping on greenish

File: pandas/io/formats/format.py (hunk 5 above)
Comment (dsaxton) [reply pandas/io/formats/format.py]: Added some types

File: pandas/io/formats/format.py (hunk 6 above)
Comment (dsaxton) [reply pandas/io/formats/format.py:1377]: It seems similar patterns occur elsewhere, although most are at the scalar level

**Comment by dsaxton** [comment]
@jreback I think this is good to go, CI failure is unrelated

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @dsaxton

**Pull request merged**